content/dev/design-pattern-observer.md

```diff
@@ -0,0 +1,235 @@
+}
+
+/**
+ * The observer interface
+ */
+interface Observer {
+    public function notify(Order $order);
+}
+{{< /highlight >}}
+
+<div class="side-note">
+    Noter que la SPL (Standard PHP Library) propose deux interfaces assez similaires : 
+
+    <ul class="fragment">
+        <li><a href="http://php.net/manual/en/class.splobserver.php" target="_blank">http://php.net/manual/en/class.splobserver.php</a></li>
+        <li><a href="http://php.net/manual/en/class.splsubject.php" target="_blank">http://php.net/manual/en/class.splsubject.php</a></li>
+    </ul>
+</div>
+
+### Le sujet observé : `CheckoutOrderHandler`
+
+Commençons par présenter une classe métier modélisant très sommairement une commande (c'est cette classe métier que manipulera notre sujet `CheckoutOrderHandler`) :
+
+{{< highlight php >}}
+<?php
+class Order {
+    const STATUS_NEW = 'new';
+    const STATUS_CHECKED_OUT = 'checked-out';
+
+    private $items;
+    private $status;
+    private $statusAt;
+
+    public function __construct() {
+        $this->items = [];
+        $this->status = self::STATUS_NEW;
+        $this->statusAt = new \DateTime();
+    }
+
+    public function checkout() {
+        $this->status = self::STATUS_CHECKED_OUT;
+        $this->statusAt = new \DateTime();
+    }
+
+    public function getStatus() { return $this->status; }
+    public function getItems() { return $this->items; }
+    public function addItem() { /* Unimplemented yet */ }
+}
+{{< /highlight >}}
+
+Voici à présent le code de notre sujet (la classe `CheckoutOrderHandler`) qui expose principalement une méthode métier `handle`.
+
+{{< highlight php >}}
+<?php
+/**
+ * The concrete subject
+ */
+class CheckoutOrderHandler implements Subject
+{
+    private $observers = [];
+
+    public function addObserver(Observer $observer) {
+        if (!in_array($observer, $this->observers)) {
+            $this->observers[] = $observer;
+        }
+    }
+    public function removeObserver(Observer $observer) { /* Unimplemented yet */ }
+
+    public function handle(Order $order) {
+        $order->checkout();
+        $this->notifyObservers($order);
+    }
+
+    public function notifyObservers(Order $order) {
+        foreach ($this->observers as $observer) {
+            $observer->notify($order);
+        }
+    }
+}
+{{< /highlight >}}
+
+* Cette classe implémente bien entendu l'interface `Subject`.
+* Elle possède en propriété la liste de ses observateurs.
+* Elle invoque sa méthode `notifyObservers` lorsqu'elle modifie l'état de la commande (`$order->checkout()`).
+
+### Les observateurs (`OrderConfirmationMailer` et `StockUpdater`)
+
+{{< highlight php >}}
+<?php
+/**
+ * A concrete observer. Service responsible for sending confirmation email to customer.
+ */
+class OrderConfirmationMailer implements Observer {
+    public function notify(Order $order) {
+        echo static::CLASS . " invoked. A confirmation mail will be sent to the customer.\n";
```

Comment:
[PSR-3](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md) ? `$this->logger->info('...');` ? :trollface: